### PR TITLE
Replace arb::util::optional with std::optional.

### LINKED ISSUE #1154: C++17 update: replace `util::optional` with `std::optional`.
Part of the C++17 update process (see task #1022).

Use `std::optional<T>` or `std::optional<std::reference_wrapper<T>>` where we currently use `util::optional<T>` and `util::optional<T&>`. Fixing up reference-usages may be a bit fiddly…

### PR DESCRIPTION
**Pull request by halfflat**
* Substitute `std::optional<T>` for `arb::util::optional<T>` for non-reference types T.
* Remove reference-deducing `util::value_by_key`; replace usages with new function `util::ptr_by_key`.
* Add some missing header includes that were required but included only transitively.
* Remove `operator<<` overload for optional in python/strprintf.hpp; replace with utility wrapper class that catches `std::optional<T>`.
* Wrap some `std::optional` values with `to_string` in ostream output in python lib.

Fixes #1154.

### CODE REVIEW

File: arbor/backends/gpu/mechanism.cpp [hunk 1]
```diff
@@ -156,7 +157,7 @@ void mechanism::instantiate(unsigned id,
     for (auto i: make_span(0, num_ions_)) {
         auto ion_binding = value_by_key(overrides.ion_rebind, ion_index_tbl[i].first).value_or(ion_index_tbl[i].first);
 
-        util::optional<ion_state&> oion = value_by_key(shared.ion_data, ion_binding);
+        ion_state* oion = value_by_key(shared.ion_data, ion_binding);
```
Comment (bcumming) [inline arbor/backends/gpu/mechanism.cpp:160]: Should this be `ptr_by_key`, not `value_by_key`?

File: arbor/mc_cell_group.cpp [hunk 2]
```diff
@@ -506,7 +506,7 @@ void mc_cell_group::remove_all_samplers() {
 std::vector<probe_metadata> mc_cell_group::get_probe_metadata(cell_member_type probe_id) const {
     // Probe associations are fixed after construction, so we do not need to grab the mutex.
 
-    util::optional<probe_tag> maybe_tag = util::value_by_key(probe_map_.tag, probe_id);
+    std::optional<probe_tag> maybe_tag = util::value_by_key(probe_map_.tag, probe_id);
```
Comment (bcumming) [inline arbor/mc_cell_group.cpp:509]: missing `#include <optional>` in this file.

File: python/cells.cpp [hunk 3]
```diff
@@ -215,7 +214,7 @@ std::string mechanism_desc_str(const arb::mechanism_desc& md) {
 
 void register_cells(pybind11::module& m) {
     using namespace pybind11::literals;
-    using arb::util::optional;
+    using std::optional;
```
Comment (bcumming) [inline python/cells.cpp:217]: missing `#include <optional>`

**Review by bcumming — CHANGES_REQUESTED** [review]
Good.
One serious comment, and a couple annoying nits about missing headers.

File: arbor/backends/gpu/mechanism.cpp (hunk 1 above)
Comment (halfflat) [reply arbor/backends/gpu/mechanism.cpp:160]: It should, yes.

File: python/cells.cpp (hunk 3 above)
Comment (halfflat) [reply python/cells.cpp:217]: No, it's up the top (line 4).

**Comment by halfflat** [comment]
bors try

**Review by bcumming — APPROVED** [review]

**Pull request merged**